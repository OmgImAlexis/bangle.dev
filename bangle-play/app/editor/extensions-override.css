.bangle-inline-suggest-emoji {
  background-color: var(--bg-stronger-color);
  color: var(--font-color);
}

.bangle-inline-suggest-emoji > div[data-is-selected='true'] {
  background-color: var(--accent-color);
}

.bangle-inline-suggest-emoji > div:hover {
  background-color: var(--accent-2-color);
}
