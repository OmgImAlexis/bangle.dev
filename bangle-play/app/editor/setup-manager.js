import { Manager } from 'bangle-plugins/collab/server/manager';
import { activeDatabaseName } from '../store/local/database-helpers';
import { defaultContent } from '../components/constants';
import { LocalDisk } from 'bangle-plugins/local-disk/local-disk';

const DEBUG = true;

export function setUpManager(db, schema) {
  const disk = new LocalDisk(db, {
    defaultDoc: defaultContent,
  });

  const manager = new Manager(schema, {
    disk,
  });

  if (
    activeDatabaseName === 'production' ||
    process.env.NODE_ENV === 'production'
  ) {
    window.addEventListener('beforeunload', (event) => {
      disk.myMainDisk.flushAll();
      event.returnValue = `Are you sure you want to leave?`;
    });
  }

  if (DEBUG) {
    window.manager = manager;
  }

  return manager;
}
