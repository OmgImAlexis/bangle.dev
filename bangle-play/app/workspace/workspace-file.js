export class WorkspaceFile {
  constructor(docName, doc, metadata = {}, opts) {
    if (!docName) {
      throw new Error('Docname needed');
    }

    if (doc === undefined) {
      throw new Error('Doc needed, though it can be null');
    }

    this._opts = opts;

    if (!opts.schema) {
      throw new Error('No schema');
    }

    this._schema = opts.schema;
    this.metadata = metadata;
    this.docName = docName;
    this.doc = doc;
    this.deleted = false;
  }

  toJSON() {
    return {
      ...this.metadata,
      _schema: 'schema',
      docName: this.docName,
      doc: this.doc,
      deleted: this.false,
    };
  }

  get title() {
    const letHydratedDoc =
      this.doc?.content && this._schema.nodeFromJSON(this.doc);
    return letHydratedDoc?.firstChild?.textContent || this.docName;
  }

  async updateDoc(doc) {}
}

export class IndexDbWorkspaceFile extends WorkspaceFile {
  static async iterateIndexDb(indexDbInstance) {
    const result = [];
    return indexDbInstance
      .iterate((value, key, iterationNumber) => {
        result.push(key);
      })
      .then(() => {
        return result;
      });
  }

  static async openFile(docName, opts) {
    const data = await opts.dbInstance.getItem(docName);
    if (data) {
      const { doc, ...metadata } = data;
      return new IndexDbWorkspaceFile(docName, doc, metadata, opts);
    }

    throw new Error('File not found');
  }

  static async getAllFilesInDb(opts) {
    const dbInstance = opts.dbInstance;
    return IndexDbWorkspaceFile.iterateIndexDb(dbInstance).then((docNames) => {
      return Promise.all(
        docNames.map((docName) => IndexDbWorkspaceFile.openFile(docName, opts)),
      );
    });
  }

  static async createFile(docName, doc, metadata, opts) {
    const data = await opts.dbInstance.getItem(docName);
    if (data) {
      throw new Error('File already exists');
    }

    const file = new IndexDbWorkspaceFile(docName, doc, metadata, opts);
    await file.updateDoc();
    return file;
  }

  static transformMetadata = (metadata) => {
    return {
      modified: 0,
      created: 0,
      version: 1,
      ...metadata,
    };
  };

  constructor(docName, doc, metadata, opts) {
    metadata = IndexDbWorkspaceFile.transformMetadata(metadata);
    super(docName, doc, metadata, opts);
    if (!opts.dbInstance) {
      throw new Error('Db Instance not found');
    }
  }

  async delete() {
    await this._opts.dbInstance.removeItem(this.docName);
    this.deleted = true;
  }

  async updateDoc(newDoc, metadata = {}) {
    if (this.deleted) {
      return;
    }

    if (newDoc) {
      this.doc = newDoc;
    }

    await this._opts.dbInstance.setItem(this.docName, {
      ...this.metadata,
      ...metadata,
      doc: this.doc,
      docName: this.docName,
    });
  }
}
