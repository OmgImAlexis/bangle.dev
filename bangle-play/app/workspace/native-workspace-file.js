import { uuid } from 'bangle-core/utils/js-utils';
import { WorkspaceFile } from './workspace-file';

export class NativeWorkspaceFile extends WorkspaceFile {
  static validateOpts(opts) {
    if (!opts.dbInstance) {
      throw new Error('Db Instance not found');
    }
    if (!opts.schema) {
      throw new Error('Db Instance not found');
    }
  }

  static async fromJSON(data, opts) {
    NativeWorkspaceFile.validateOpts(opts);
    const { doc, docName, metadata } = data;
    if (!data.doc || !data.docName) {
      throw new Error('Doc and docName are necessary');
    }
    return NativeWorkspaceFile.createFile(docName, doc, metadata, opts);
  }

  static async openFile(docName, opts) {
    NativeWorkspaceFile.validateOpts(opts);

    const data = await opts.dbInstance.getItem(docName);
    if (data) {
      const { doc, ...metadata } = data;
      return new NativeWorkspaceFile(docName, doc, metadata, opts);
    }

    throw new Error('File not found');
  }

  static async getAllFilesInDb(opts) {
    NativeWorkspaceFile.validateOpts(opts);

    const dbInstance = opts.dbInstance;
    return iterateIndexDb(dbInstance).then((docNames) => {
      return Promise.all(
        docNames.map((docName) => NativeWorkspaceFile.openFile(docName, opts)),
      );
    });
  }

  static async createFile(suggestedDocName, doc, metadata, opts) {
    NativeWorkspaceFile.validateOpts(opts);

    const docName = opts.dbInstance.createNewItemKey(suggestedDocName);

    const data = await opts.dbInstance.getItem(docName);
    if (data) {
      throw new Error('File already exists');
    }

    const file = new NativeWorkspaceFile(docName, doc, metadata, opts);
    await file.updateDoc();
    return file;
  }

  _dbInstance = null;

  constructor(docName, doc, metadata, opts) {
    if (!docName.endsWith('.md')) {
      throw new Error('Only md files');
    }
    metadata = {
      created: Date.now(),
      ...metadata,
      modified: Date.now(),
    };

    super(docName, doc, metadata, opts);
    NativeWorkspaceFile.validateOpts(opts);
    this._dbInstance = opts.dbInstance;
  }

  async delete() {
    super.delete();
    await this._dbInstance.removeItem(this.docName);
  }

  async updateDoc(newDoc, metadata = {}) {
    if (this.deleted) {
      return;
    }

    if (newDoc) {
      this.doc = newDoc;
    }

    await this._dbInstance.setItem(this.docName, {
      doc: this.doc,
      docName: this.docName,
      metadata,
    });
  }
}

function iterateIndexDb(indexDbInstance) {
  const result = [];
  return indexDbInstance
    .iterate((value, key, iterationNumber) => {
      result.push(key);
    })
    .then(() => {
      return result;
    });
}
