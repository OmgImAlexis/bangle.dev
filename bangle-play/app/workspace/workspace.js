import { IndexDbWorkspaceFile } from './workspace-file';
const LOG = false;

let log = LOG ? console.log.bind(console, 'play/workspace') : () => {};

// TODO if a file fails to open for whatever reason donot fail
// allow user to open another file
export class Workspace {
  /**
   * @type {WorkspaceFile[]}
   */
  files;

  /**
   * @returns {Promise<WorkspaceFile[]>>}
   */

  constructor(name, files, opts) {
    this.name = name;
    this._opts = opts;
    if (!opts.schema) {
      throw new Error('Schema needed');
    }

    this.files = files.sort((a, b) => a.docName.localeCompare(b.docName));
  }

  /**
   * @returns {Promise<WorkspaceFile>} docName
   */
  getFile(docName) {
    return this.files.find((file) => file.docName === docName);
  }

  hasFile(docName) {
    return Boolean(this.getFile(docName));
  }

  linkFile(file) {
    return new IndexDbWorkspace(this.name, [...this.files, file], this._opts);
  }

  unlinkFile(file) {
    return new IndexDbWorkspace(
      this.name,
      this.files.filter((f) => f.docName !== file.docName),
      this._opts,
    );
  }
}

export class IndexDbWorkspace extends Workspace {
  static async create(name, opts) {
    let files = await IndexDbWorkspaceFile.getAllFilesInDb(opts);
    return new IndexDbWorkspace(name, files, opts);
  }

  /**
   *
   * @param {*} payload
   * @returns {Promise<IndexDbWorkspaceFile>}
   */
  async createFile(docName, doc) {
    return IndexDbWorkspaceFile.createFile(docName, doc, undefined, this._opts);
  }
}
