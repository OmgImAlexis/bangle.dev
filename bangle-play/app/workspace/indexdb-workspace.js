import { getIdleCallback, uuid } from 'bangle-core/utils/js-utils';
import localforage from 'localforage';
import { Workspace } from './workspace';
import { IndexDbWorkspaceFile } from './workspace-file';
const LOG = true;

let log = LOG ? console.log.bind(console, 'play/idb-workspace') : () => {};

export const TYPE_INDEXDB = 'indexdb';
const createIndexdbUid = () => TYPE_INDEXDB + '_' + uuid(6);

export class IndexDbWorkspace extends Workspace {
  static getDbInstance = async (uid, metadata, schema) => {
    if (!metadata) {
      throw new Error('metadata needed');
    }
    if (!schema) {
      throw new Error('schema needed');
    }

    return localforage.createInstance({
      name: uid,
    });
  };

  static async restoreWorkspaceFromBackupFile(data, schema) {
    const uid = createIndexdbUid();
    const dbInstance = await IndexDbWorkspace.getDbInstance(uid, {}, schema);

    let { name, files, metadata } = data;

    const type = TYPE_INDEXDB;

    // old style backup
    if (Array.isArray(data)) {
      name = 'pyare-mohan' + Math.floor(100 * Math.random());
      files = await Promise.all(
        data.map((item) =>
          IndexDbWorkspaceFile.fromJSON(item, { schema, dbInstance }),
        ),
      );
      metadata = {};
    } else {
      files = await Promise.all(
        files.map((item) =>
          IndexDbWorkspaceFile.fromJSON(item, { schema, dbInstance }),
        ),
      );
    }

    const opts = {
      dbInstance,
      schema,
      metadata,
      name,
    };

    const instance = new IndexDbWorkspace(uid, files, type, opts);
    await instance.persistWorkspaceInfo();
    return instance;
  }

  static async openWorkspace(uid, name, schema, metadata = {}) {
    const dbInstance = await IndexDbWorkspace.getDbInstance(
      uid,
      metadata,
      schema,
    );
    const opts = {
      name,
      dbInstance,
      schema,
      metadata,
    };
    let files = await IndexDbWorkspaceFile.getAllFilesInDb(opts);
    const instance = new IndexDbWorkspace(uid, files, TYPE_INDEXDB, opts);
    await instance.persistWorkspaceInfo();
    return instance;
  }

  static openExistingWorkspace(workspaceInfo, schema) {
    const { uid, name, metadata } = workspaceInfo;

    return IndexDbWorkspace.openWorkspace(uid, name, schema, metadata);
  }

  static async createWorkspace(name, schema) {
    const uid = createIndexdbUid();
    return IndexDbWorkspace.openWorkspace(uid, name, schema);
  }

  updateFiles(files) {
    return new IndexDbWorkspace(this.uid, files, this.type, this._opts);
  }

  /**
   * TODO remove this, no need to put it in here
   */
  async createFile(docName, doc) {
    return IndexDbWorkspaceFile.createFile(docName, doc, undefined, this._opts);
  }

  async deleteWorkspace() {
    if (this.deleted) {
      return;
    }

    super.deleteWorkspace();
    let files = await IndexDbWorkspaceFile.getAllFilesInDb(this._opts);
    files.forEach((value) => {
      value.delete();
    });
    getIdleCallback(() => {
      localforage.dropInstance({ name: this.uid });
    });
  }
}
