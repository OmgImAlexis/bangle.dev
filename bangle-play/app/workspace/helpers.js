import { NativeWorkspace, TYPE_NATIVE } from './native-workspace';
import { IndexDbWorkspace, TYPE_INDEXDB } from './indexdb-workspace';

export function openExistingWorkspace(workspaceInfo, schema) {
  if (workspaceInfo.uid.startsWith(TYPE_INDEXDB)) {
    return IndexDbWorkspace.openExistingWorkspace(workspaceInfo, schema);
  }

  if (workspaceInfo.uid.startsWith(TYPE_NATIVE)) {
    return NativeWorkspace.openExistingWorkspace(workspaceInfo, schema);
  }

  throw new Error('Not implemented ' + workspaceInfo.uid);
}

export function restoreWorkspaceFromBackupFile(schema, data, type) {
  if (type === TYPE_INDEXDB) {
    return IndexDbWorkspace.restoreWorkspaceFromBackupFile(data, schema);
  }

  if (type === TYPE_NATIVE) {
    return NativeWorkspace.restoreWorkspaceFromBackupFile(data, schema);
  }

  throw new Error('Not implemented ' + type);
}
