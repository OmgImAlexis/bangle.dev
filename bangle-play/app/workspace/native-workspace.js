import localforage from 'localforage';

import { Workspace } from './workspace';
import { getIdleCallback, sleep, uuid } from 'bangle-core/utils/js-utils';
import { NativeWorkspaceFile } from './native-workspace-file';
import {
  readFile,
  recurseDirHandle,
} from '../components/native-file-api/helper';
import { markdownParser } from 'bangle-plugins/markdown/markdown-parser';
import { markdownSerializer } from 'bangle-plugins/markdown/markdown-serializer';
const LOG = true;

let log = LOG ? console.log.bind(console, 'play/native-workspace') : () => {};

export const TYPE_NATIVE = 'native';
const createNativeUID = () => TYPE_NATIVE + '_' + uuid(6);

export class NativeWorkspace extends Workspace {
  static async getDbInstance(uid, metadata, schema) {
    if (!metadata) {
      throw new Error('metadata needed');
    }
    if (!schema) {
      throw new Error('schema needed');
    }

    return FSStorage.createInstance(metadata.dirHandle, schema);
  }

  static async restoreWorkspaceFromBackupFile(data, schema) {
    const uid = createNativeUID();
    const dbInstance = await NativeWorkspace.getDbInstance(uid, {}, schema);

    let { name, files, metadata } = data;

    // old style backup
    if (Array.isArray(data)) {
      name = 'pyare-mohan' + Math.floor(100 * Math.random());
      files = await Promise.all(
        data.map((item) =>
          NativeWorkspaceFile.fromJSON(item, { schema, dbInstance }),
        ),
      );
      metadata = {};
    } else {
      files = await Promise.all(
        files.map((item) =>
          NativeWorkspaceFile.fromJSON(item, { schema, dbInstance }),
        ),
      );
    }

    const opts = {
      dbInstance,
      schema,
      metadata,
      name,
    };

    const type = TYPE_NATIVE;
    const instance = new NativeWorkspace(uid, files, type, opts);
    await instance.persistWorkspaceInfo();
    return instance;
  }

  static async openWorkspace(uid, name, schema, metadata = {}) {
    const dbInstance = await NativeWorkspace.getDbInstance(
      uid,
      metadata,
      schema,
    );
    const opts = {
      name,
      dbInstance,
      schema,
      metadata,
    };
    let files = await NativeWorkspaceFile.getAllFilesInDb(opts);
    const instance = new NativeWorkspace(uid, files, TYPE_NATIVE, opts);
    await instance.persistWorkspaceInfo();
    return instance;
  }

  static openExistingWorkspace(workspaceInfo, schema) {
    const { uid, name, metadata } = workspaceInfo;
    return NativeWorkspace.openWorkspace(uid, name, schema, metadata);
  }

  static async createWorkspace(name, schema) {
    const uid = createNativeUID();
    return NativeWorkspace.openWorkspace(uid, name, schema);
  }

  updateFiles(files) {
    return new NativeWorkspace(this.uid, files, this.type, this._opts);
  }

  /**
   * TODO remove this, no need to put it in here
   */
  async createFile(docName, doc) {
    return NativeWorkspaceFile.createFile(docName, doc, undefined, this._opts);
  }

  async deleteWorkspace() {
    if (this.deleted) {
      return;
    }

    super.deleteWorkspace();
    let files = await NativeWorkspaceFile.getAllFilesInDb(this._opts);
    files.forEach((value) => {
      value.delete();
    });
    getIdleCallback(() => {
      dropInstance({ name: this.uid });
    });
  }
}

export async function pickADirectory(dirHandle) {
  if (dirHandle) {
    let permission = await verifyPermission(dirHandle);
    if (!permission) {
      throw new NoPermissionError(
        'The permission to edit directory was denied',
      );
    }
  } else {
    try {
      dirHandle = await window.showDirectoryPicker();
    } catch (err) {
      console.error(err);
      throw new NoPermissionError(err.message);
    }
  }

  return dirHandle;
}

async function verifyPermission(fileHandle, withWrite = true) {
  const opts = {};
  if (withWrite) {
    opts.writable = true;
    // For Chrome 86 and later...
    opts.mode = 'readwrite';
  }
  // Check if we already have permission, if so, return true.
  if ((await fileHandle.queryPermission(opts)) === 'granted') {
    return true;
  }
  // Request permission to the file, if the user grants permission, return true.
  if ((await fileHandle.requestPermission(opts)) === 'granted') {
    return true;
  }
  // The user did nt grant permission, return false.
  return false;
}

export class NoPermissionError extends Error {
  constructor(errorCode, body) {
    super(body);
    this.errorCode = errorCode;
    this.body = body;
    Error.captureStackTrace(this, NoPermissionError);
    this.name = this.constructor.name;
  }
}

class FSStorage {
  // TODO check if the new directory is something we already have
  // in our database
  static async createInstance(dirHandle, schema) {
    await pickADirectory(dirHandle);
    return new FSStorage(dirHandle, schema);
  }

  static getFilePathKey = (filePathHandles) =>
    filePathHandles.map((r) => r.name).join('/');

  _transformer = async (filePathHandles) => {
    const textContent = await readFile(
      await getLast(filePathHandles).getFile(),
    );
    const key = FSStorage.getFilePathKey(filePathHandles);
    const doc = await this._parser.parse(textContent);
    return {
      key,
      value: doc.toJSON(),
    };
  };

  _updateFilePathHandles = async () => {
    this._filePathHandles = await recurseDirHandle(this._rootDirHandle, {
      allowedFile: (entry) => entry.name.endsWith('.md'),
    });
    return this._filePathHandles;
  };

  _findPathHandlersByKey = (key) => {
    return this._filePathHandles.find(
      (paths) => FSStorage.getFilePathKey(paths) === key,
    );
  };

  constructor(dirHandle, schema) {
    this._rootDirHandle = dirHandle;
    this._schema = schema;

    this._parser = markdownParser(schema);
    this._serializer = markdownSerializer(schema);
  }

  async iterate(cb) {
    await this._updateFilePathHandles();

    await Promise.all(
      this._filePathHandles.map(async (filePathHandle, i) => {
        // const fileHandle = getLast(filePathHandle);
        const key = FSStorage.getFilePathKey(filePathHandle);
        cb(null, key, i);
      }),
    );
  }

  // TODO support created nested keys
  createNewItemKey(fileName = uuid(6), parent) {
    // root
    if (!parent) {
      return [this._rootDirHandle.name, fileName + '.md'].join('/');
    }

    throw new Error('Not implemented');
  }

  async getItem(key) {
    const match = this._findPathHandlersByKey(key);

    if (match) {
      const { value } = await this._transformer(match);
      return { doc: value };
    }
  }

  async removeItem(key) {
    const match = this._findPathHandlersByKey(key);
    if (!match) {
      throw new Error(`Key ${key} not found`);
    }

    await match[match.length - 2].removeEntry(getLast(match).name);
    await this._updateFilePathHandles();
  }

  async setItem(key, value) {
    const _setItem = async (match) => {
      const handler = match[match.length - 1];
      let data = '';
      // When a new file is created initially the value.doc is null
      if (value.doc) {
        const doc = this._schema.nodeFromJSON(value.doc);
        data = this._serializer.serialize(doc);
      }
      await writeFile(handler, data);
      return;
    };

    let match = this._filePathHandles.find(
      (paths) => FSStorage.getFilePathKey(paths) === key,
    );

    if (match) {
      await _setItem(match);
      return;
    }

    // TODO currently this only creates rootlevel files
    // if not found attempt to create it
    const name = getLast(key.split('/'));
    await this._rootDirHandle.getFileHandle(name, { create: true });
    await this._updateFilePathHandles();
    match = this._filePathHandles.find(
      (paths) => FSStorage.getFilePathKey(paths) === key,
    );
    await _setItem(match);
    return;
  }
}

async function writeFile(fileHandle, contents) {
  // Support for Chrome 82 and earlier.
  if (fileHandle.createWriter) {
    // Create a writer (request permission if necessary).
    const writer = await fileHandle.createWriter();
    // Write the full length of the contents
    await writer.write(0, contents);
    // Close the file and write the contents to disk
    await writer.close();
    return;
  }
  // For Chrome 83 and later.
  // Create a FileSystemWritableFileStream to write to.
  const writable = await fileHandle.createWritable();
  // Write the contents of the file to the stream.
  await writable.write(contents);
  // Close the file and write the contents to disk.
  await writable.close();
}

function getLast(array) {
  return array[array.length - 1];
}
