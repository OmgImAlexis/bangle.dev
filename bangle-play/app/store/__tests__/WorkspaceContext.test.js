import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { extensions } from 'bangle-play/app/editor/extensions';
import { getSchema } from 'bangle-play/app/editor/utils';
import { IndexDbWorkspace } from 'bangle-play/app/workspace/workspace';
import { IndexDbWorkspaceFile } from 'bangle-play/app/workspace/workspace-file';

import localforage from 'localforage';
import {
  workspaceActions,
  WorkspaceContext,
  WorkspaceContextProvider,
} from '../WorkspaceContext';

jest.mock('localforage', () => {
  const instance = {
    iterate: jest.fn(async () => {}),
    setItem: jest.fn(async () => {}),
    removeItem: jest.fn(async () => {}),
    getItem: jest.fn(async () => {}),
  };

  return {
    config: jest.fn(),
    createInstance: jest.fn(({ name } = {}) => {
      if (name === 'workspaces/1') {
        return {
          iterate: jest.fn(async () => {}),
          setItem: jest.fn(async () => {}),
          removeItem: jest.fn(async () => {}),
          getItem: jest.fn(async () => {}),
        };
      }
      return instance;
    }),
  };
});
const DateNowBackup = jest.fn();
describe('index db workspace', () => {
  let dbInstance;
  const schema = getSchema(extensions());
  beforeEach(async () => {
    Date.now = jest.fn(() => 1);
    dbInstance = localforage.createInstance();
    dbInstance.getItem = jest.fn(async (docName) => ({ docName, doc: null }));
    dbInstance.iterate.mockImplementationOnce(async (cb) => {
      Array.from({ length: 5 }, (_, k) =>
        cb(
          {
            docName: k + '',
            doc: null,
          },
          k + '',
          k,
        ),
      );
    });
  });
  afterEach(() => {
    Date.now = DateNowBackup;
  });
  test('Creates workspace correctly', async () => {
    await IndexDbWorkspace.createWorkspace('test_db', schema);
  });

  test('Adds files correctly', async () => {
    const workspace = await IndexDbWorkspace.createWorkspace('test_db', schema);

    expect(workspace.files).toMatchSnapshot();
  });

  test('getFile ', async () => {
    const workspace = await IndexDbWorkspace.createWorkspace('test_db', schema);

    expect(workspace.getFile('1')).toMatchInlineSnapshot(`
      Object {
        "doc": null,
        "docName": "1",
        "metadata": Object {
          "created": 1,
          "docName": "1",
          "modified": 1,
        },
      }
    `);
  });

  test('hasFile ', async () => {
    const workspace = await IndexDbWorkspace.createWorkspace('test_db', schema);

    expect(workspace.hasFile('2')).toBe(true);
  });

  test('link file', async () => {
    const workspace = await IndexDbWorkspace.createWorkspace('test_db', schema);

    const newFile = new IndexDbWorkspaceFile(
      'test_doc',
      null,
      {},
      { dbInstance: dbInstance, schema: schema },
    );

    let newWorkspace = workspace.linkFile(newFile);
    expect(workspace.hasFile('test_doc')).toBe(false);
    expect(newWorkspace.hasFile('test_doc')).toBe(true);
  });

  test('unlink file', async () => {
    const workspace = await IndexDbWorkspace.createWorkspace('test_db', schema);

    let newWorkspace = workspace.unlinkFile(workspace.getFile('2'));

    expect(workspace.hasFile('2')).toBe(true);
    expect(newWorkspace.hasFile('2')).toBe(false);
  });
});

describe('indexdb workspaceContext', () => {
  let dbInstance;
  const customRender = (child, { ...renderOptions } = {}) => {
    return render(
      <WorkspaceContextProvider>{child}</WorkspaceContextProvider>,
      renderOptions,
    );
  };

  beforeEach(async () => {
    Date.now = jest.fn(() => 1);
  });
  afterEach(() => {
    Date.now = DateNowBackup;
  });

  beforeEach(async () => {
    dbInstance = localforage.createInstance();
    dbInstance.getItem = jest.fn(async (docName) => ({ docName, doc: null }));
    dbInstance.iterate.mockImplementation(async (cb) => {
      Array.from({ length: 5 }, (_, k) =>
        cb(
          {
            docName: k + '',
            doc: null,
          },
          k + '',
          k,
        ),
      );
    });
  });

  test('sets up workspace', async () => {
    customRender(
      <WorkspaceContext.Consumer>
        {({ workspace }) =>
          workspace && (
            <span data-testid="result">
              Result: {workspace.files.map((f) => f.docName)}
            </span>
          )
        }
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    expect(screen.getByTestId('result').textContent).toMatchInlineSnapshot(
      `"Result: 01234"`,
    );
    expect(dbInstance.getItem).toBeCalledTimes(5);
  });

  test('has updateContext', async () => {
    customRender(
      <WorkspaceContext.Consumer>
        {({ updateContext } = {}) =>
          updateContext && (
            <span data-testid="result">{Boolean(updateContext)}</span>
          )
        }
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    expect(screen.getByTestId('result')).toBeTruthy();
  });

  test('has one openedDocuments initially', async () => {
    customRender(
      <WorkspaceContext.Consumer>
        {({ workspace, openedDocuments } = {}) => {
          if (!workspace) {
            return;
          }
          expect(openedDocuments.every((r) => r.docName)).toBe(true);
          expect(openedDocuments.every((r) => r.key)).toBe(true);
          return (
            <span data-testid="result">
              {openedDocuments.map((r) => r.docName)}
            </span>
          );
        }}
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    expect(screen.getByTestId('result')).toMatchInlineSnapshot(`
      <span
        data-testid="result"
      >
        0
      </span>
    `);
  });

  test('openWorkspaceFile', async () => {
    let _updateContext;
    customRender(
      <WorkspaceContext.Consumer>
        {({ workspace, openedDocuments, updateContext } = {}) => {
          if (!workspace) {
            return;
          }
          _updateContext = updateContext;
          return (
            <span data-testid="result">
              {openedDocuments.map((r) => r.docName)}
            </span>
          );
        }}
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    await _updateContext(workspaceActions.openWorkspaceFile('2'));

    expect(screen.getByTestId('result')).toMatchInlineSnapshot(`
      <span
        data-testid="result"
      >
        2
        0
      </span>
    `);

    // opening again should work
    await _updateContext(workspaceActions.openWorkspaceFile('2'));

    expect(screen.getByTestId('result').textContent).toMatchInlineSnapshot(
      `"22"`,
    );
  });

  test('openBlankWorkspaceFile', async () => {
    let _updateContext;
    let _openedDocuments;
    let _workspace;

    dbInstance.iterate.mockImplementation(async (cb) => {
      Array.from({ length: 2 }, (_, k) =>
        cb(
          {
            docName: k + '',
            doc: null,
          },
          k + '',
          k,
        ),
      );
    });

    dbInstance.getItem = jest.fn(async (docName) => {
      // we only give out 2 files in mock iterate
      if (['0', '1'].includes(docName)) {
        return {
          docName,
          doc: null,
        };
      }
    });

    customRender(
      <WorkspaceContext.Consumer>
        {({ workspace, openedDocuments, updateContext } = {}) => {
          if (!workspace) {
            return;
          }
          _workspace = workspace;
          _updateContext = updateContext;
          _openedDocuments = openedDocuments;
          return (
            <span data-testid="result">
              {openedDocuments.map((r) => r.docName)}
            </span>
          );
        }}
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    await _updateContext(workspaceActions.openBlankWorkspaceFile());

    expect(_openedDocuments.length).toBe(2);
    // its a random string
    expect(typeof _openedDocuments[0].docName === 'string').toBe(true);

    expect(_workspace.hasFile(_openedDocuments[0].docName)).toBe(true);
    expect(dbInstance.setItem).toHaveBeenNthCalledWith(
      1,
      _openedDocuments[0].docName,
      expect.any(Object),
    );
  });

  test('deleteWorkspaceFile', async () => {
    let _updateContext;
    let _openedDocuments;
    let _workspace;
    customRender(
      <WorkspaceContext.Consumer>
        {({ workspace, openedDocuments, updateContext } = {}) => {
          if (!workspace) {
            return;
          }
          _workspace = workspace;
          _updateContext = updateContext;
          _openedDocuments = openedDocuments;
          return (
            <span data-testid="result">
              {openedDocuments.map((r) => r.docName)}
            </span>
          );
        }}
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    await _updateContext(workspaceActions.openWorkspaceFile('2'));
    expect(_openedDocuments[0].docName).toBe('2');

    await _updateContext(workspaceActions.deleteWorkspaceFile('2'));

    expect(_openedDocuments.length).toBe(1);
    // its a random string
    expect(_openedDocuments[0].docName).toBe('0');

    expect(_workspace.hasFile(_openedDocuments[0].docName)).toBe(true);
    expect(dbInstance.removeItem).toBeCalledTimes(1);
    expect(dbInstance.removeItem).toHaveBeenNthCalledWith(1, '2');
  });
});
