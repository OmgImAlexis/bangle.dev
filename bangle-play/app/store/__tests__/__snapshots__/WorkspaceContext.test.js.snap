// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index db workspace Adds files correctly 1`] = `
Array [
  Object {
    "_schema": "schema",
    "created": 0,
    "deleted": undefined,
    "doc": null,
    "docName": "0",
    "modified": 0,
    "version": 1,
  },
  Object {
    "_schema": "schema",
    "created": 0,
    "deleted": undefined,
    "doc": null,
    "docName": "1",
    "modified": 0,
    "version": 1,
  },
  Object {
    "_schema": "schema",
    "created": 0,
    "deleted": undefined,
    "doc": null,
    "docName": "2",
    "modified": 0,
    "version": 1,
  },
  Object {
    "_schema": "schema",
    "created": 0,
    "deleted": undefined,
    "doc": null,
    "docName": "3",
    "modified": 0,
    "version": 1,
  },
  Object {
    "_schema": "schema",
    "created": 0,
    "deleted": undefined,
    "doc": null,
    "docName": "4",
    "modified": 0,
    "version": 1,
  },
]
`;
